Add a pref in order to cap the canvas area to a factor of the window one (bug 1958015)

calixteman opened the pull request:
This way it helps to reduce the overall canvas dimensions and make the rendering faster. The drawback is that when scrolling, the page can be blurry in waiting for the rendering.

The default value is 200% on desktop and will be 100% for GeckoView.

Snuffleupagus commented:
This limit really seems *way too low* to be practical/helpful, since I cannot even open e.g. the `tracemonkey.pdf` document at the auto-zoom level (i.e. 125%) without triggering use of detailViews which is quite annoying when scrolling through the document (since there's now lots of re-rendering happening all the time).

marco-c commented:
What if we make the value dynamic based on the same factors we use for the auto-zoom level? If the text is very small, it's very likely the user will zoom and scroll, and so we might want to pre-render more.

marco-c commented:
Another idea: we could render with a smaller canvas, but pre-render more area around it (or even simply a larger canvas) while on idle.

Snuffleupagus commented:
Given that the bug now links to a GeckoView-specific bug, I'd be more OK with the general idea of this patch *if we use pre-processor statements to limit this functionality to the GV-viewer* (in `web/app_options.js`) since this feels "unnecessary" on desktop.

Also, is this intended as a temporary solution until issue #6419 is fully fixed?

marco-c commented:
We were thinking of three things:
1. finish up https://github.com/mozilla/pdf.js/pull/19043;
2. support drawing in a worker;
3. https://github.com/mozilla/pdf.js/issues/6419 (which can then benefit from drawing in parallel given 2).

calixteman commented:
It isn't only improving speed for GV but for Desktop too when  the pdf is large.
For example, with `wuppertal_2012.pdf` it takes around 18s to render at 210% with this patch when it takes several minutes without.
Maybe we can try to compute a better value when creating the first canvas and based on the time used to draw everything.
Something else which could improve the rendering of large pdfs like maps, would be to not draw what it isn't in the detail view.
For example wherever an image is we rescale it, but it's outside the view then it's useless, the same when drawing a group (and if it's partially rendered we should set a clip box), the same when stroking where we've to compute some scale factors, ...
If we implement tiling we'd have to do that so it's probably a good idea to start asap.

Snuffleupagus commented on web/pdf_page_detail_view.js:
```diff
@@ -140,10 +141,21 @@ class PDFPageDetailView extends BasePDFPageView {
     }
 
-    const { viewport, maxCanvasPixels } = this.pageView;
+    const { viewport } = this.pageView;
```
Can't we just do the following here, and avoid any changes in the constructor?
```suggestion
    const { viewport, capCanvasAreaFactor } = this.pageView;
```

Snuffleupagus commented on src/display/display_utils.js:
```diff
@@ -658,11 +658,21 @@ class OutputScale {
 
+    if (capCanvasAreaFactor >= 0) {
+      const { innerWidth, innerHeight } = window;
+      maxPixels = Math.min(
+        maxPixels,
+        innerWidth *
+          innerHeight *
+          (1 + capCanvasAreaFactor / 100) *
+          OutputScale.pixelRatio ** 2
+      );
```
Sorry, but this really won't work if `maxPixels` is either `-1` or `0` both of which are possible/valid values for that parameter (note the condition just below).
```suggestion
      const capPixels = 
        innerWidth *
          innerHeight *
          (1 + capCanvasAreaFactor / 100) *
          OutputScale.pixelRatio ** 2;

      maxPixels = maxPixels > 0 ? Math.min(maxPixels, capPixels) : capPixels;
```

Snuffleupagus commented on src/display/display_utils.js:
```diff
@@ -658,11 +658,21 @@ class OutputScale {
    *   `false` otherwise.
    */
-  limitCanvas(width, height, maxPixels, maxDim) {
+  limitCanvas(width, height, maxPixels, maxDim, capCanvasAreaFactor) {
```
Nit: The other parameters don't contain the word "Canvas" to shorten the method signature a little bit.

Also, since the thumbnail rendering also uses this method (where the cap-parameter won't be used) let's give the new parameter a default value.
```suggestion
  limitCanvas(width, height, maxPixels, maxDim, capAreaFactor = -1) {
```

Snuffleupagus reviewed (changes requested):
Since the pixel-limits now also depend on the window size, what happens if the viewer is resized such that it becomes bigger: will re-rendering be triggered in that case (since I believe that it should)?

calixteman committed:
Add a pref in order to cap the canvas area to a factor of the window one (bug 1958015)

This way it helps to reduce the overall canvas dimensions and make the rendering faster.
The drawback is that when scrolling, the page can be blurry in waiting for the rendering.

The default value is 200% on desktop and will be 100% for GeckoView.

calixteman commented:
For information, I profiled wuppertal_2012.pdf at 300% on desktop with a local dev server (Windows 11) and with a local build of Fenix on a real pixel 7 phone.
Here are the results (time spent in `executeOperatorList` in seconds);
Cap factor     Desktop    Fenix
-1                  5.8             5.7
0                   3.3             3.3
100               4.3             3.7
200               5.0             4.2
300               5.7             4.4

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -543,8 +555,49 @@ describe("PDF viewer", () => {
+              expect(after.length)
+                .withContext(`In ${browserName}, after`)
+                .toBe(2);
+              expect(after[0].width)
+                .withContext(`In ${browserName}`)
+                .toBe(582 * pixelRatio);
+              expect(after[0].height)
+                .withContext(`In ${browserName}`)
+                .toBe(823 * pixelRatio);
+              expect(after[1].width)
+                .withContext(`In ${browserName}`)
+                .toBeLessThan(810 * pixelRatio);
```
Where are these hard-coded factors coming from, and is there a risk that they might lead to intermittent errors?

In any case, a comment explaining these factors seem like a good idea.

Snuffleupagus commented on web/app.js:
```diff
@@ -362,6 +362,7 @@ const PDFViewerApplication = {
     if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
       Object.assign(opts, {
         docBaseUrl: x => x,
+        capCanvasAreaFactor: x => parseInt(x),
```
Nit: Alphabetical order please.
```suggestion
        capCanvasAreaFactor: x => parseInt(x),
        docBaseUrl: x => x,
```

Snuffleupagus commented on web/app_options.js:
```diff
@@ -168,6 +168,11 @@ const defaultOptions = {
   },
+  capCanvasAreaFactor: {
+    /** @type {number} */
+    value: 200,
```
Given that https://github.com/mozilla/pdf.js/pull/19755#issue-2967184930 now contains:
> The default value is 200% on desktop and will be 100% for GeckoView.

I'd expect that to be reflected in the definition here?

```suggestion
    value: (
      typeof PDFJSDev === "undefined"
        ? window.isGECKOVIEW
        : PDFJSDev.test("GECKOVIEW")
    )
      ? 100
      : 200,
```

calixteman replied:
We've only one pref file in m-c, so we need to update `PdfJsOverridePrefs.js`, right ? or am I missing something ?

Snuffleupagus replied:
Right, we don't have separate prefs-files for desktop/geckoview; I suppose that we can disregard the suggestion :-)

Snuffleupagus reviewed (changes requested):
It doesn't appear that the question from https://github.com/mozilla/pdf.js/pull/19755#pullrequestreview-2764637965 has been addressed, since it still seems to me that if the window size is increased we should potentially re-render?

One way to handle that may be to keep track of the last value computed by `getCappedWindowArea(...)` somewhere in the viewer, and if it increased then call `pdfViewer.refresh()` instead of https://github.com/mozilla/pdf.js/blob/6f052312d625224173db36d3e661657a89cf1865/web/app.js#L2483

calixteman commented:
/botio test

Snuffleupagus commented on src/display/display_utils.js:
```diff
@@ -685,6 +692,23 @@ class OutputScale {
+
+  static getCappedWindowArea(capAreaFactor) {
+    if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
+      return (
+        window.innerWidth *
+        window.innerHeight *
+        this.pixelRatio ** 2 *
+        (1 + capAreaFactor / 100)
+      );
+    }
+    return (
+      window.screen.availWidth *
+      window.screen.availHeight *
+      this.pixelRatio ** 2 *
+      (1 + capAreaFactor / 100)
+    );
```
It just occurred to me that the pixelRatio can be a decimal number and that the capAreaFactor could be any integer (e.g. `123`), hence the resulting value could also become a decimal number which seems a bit strange.
While that probably won't cause any issues, it still seems nice to ensure that the result is always an integer.

```suggestion
      return Math.ceil(
        window.innerWidth *
        window.innerHeight *
        this.pixelRatio ** 2 *
        (1 + capAreaFactor / 100)
      );
    }
    return Math.ceil(
      window.screen.availWidth *
      window.screen.availHeight *
      this.pixelRatio ** 2 *
      (1 + capAreaFactor / 100)
    );
```

Snuffleupagus commented on web/pdf_page_view.js:
```diff
@@ -188,6 +191,8 @@ class PDFPageView extends BasePDFPageView {
       options.maxCanvasPixels ?? AppOptions.get("maxCanvasPixels");
     this.maxCanvasDim = options.maxCanvasDim || AppOptions.get("maxCanvasDim");
+    this.capCanvasAreaFactor =
+      options.capCanvasAreaFactor || AppOptions.get("capCanvasAreaFactor");
```
Is `0` a valid value for the option/preference, because if so the fallback should probably be tweaked a little?
```suggestion
      options.capCanvasAreaFactor ?? AppOptions.get("capCanvasAreaFactor");
```

Snuffleupagus reviewed (approved):
There's a small typo in the commit message: `areato` -> `area to`.

r=me, with a couple more comments; thank you.

calixteman commented:
/botio integrationtest
